nrf: Change UART to be non-blocking and use IRQs

dpgeorge opened the pull request:
This change adds an internal ring buffer to the UART to store incoming characters, makes it non-blocking, and allows ctrl-C (via UART REPL) to asynchronously interrupt a running program.

@glennrub I'm not sure I've done this properly, so please can you take a look.  Also, maybe newer versions of nrfx handle non-blocking UART differently?

dpgeorge committed:
nrf/uart: Change UART driver to be non-blocking and use IRQs.

dpgeorge committed:
nrf/mpconfigport.h: Enable MICROPY_KBD_EXCEPTION by default.

glennrub commented on ports/nrf/modules/machine/uart.c:
```diff
@@ -182,8 +202,10 @@ STATIC mp_obj_t machine_hard_uart_make_new(const mp_obj_type_t *type, size_t n_a
+    nrfx_uart_init(self->p_uart, &config, m_rx_handler);
+    rx_ringbuf.iput = rx_ringbuf.iget = 0;
+    tx_pending = false;
+    nrfx_uart_rx(self->p_uart, &rx_buf[0], 1);
```
I'm a bit unsure about this one, and it might be an edge case. The STARTRX task in the UART peripheral is issued in next line, this means that if there is a byte sent to UART as this point, it might not trigger an IRQ. So, if this is the trigger to get recursive RX IRQ's, maybe we should move this after RX has been enabled to be safe? I could check a bit up with the specification if the peripheral will trigger a IRQ on outstanding RX bytes.

dpgeorge replied:
I'm not sure swapping the lines will make a difference because `rx_enable()` is anyway called by `nrfx_uart_rx()` if RX is not already enabled (see the check `if ((!p_cb->rx_enabled) && (!second_buffer))`).  So I guess we just remove the following `nrfx_uart_rx_enable()` altogether?

glennrub replied:
At a first glance this looked correct to me, that we could remove the `nrfx_uart_rx_enable()` as the `rx_enable()` is done internally. 

However, looking over the code again, to me it looks like we have to call the `nrfx_uart_rx_enable()` once to get `p_cb->rx_enabled = true`. If not enabled (permanently false), `rx_enable()` will be called for all subsequent calls to `nrfx_uart_rx()`, which also indirectly clear the UART peripheral `NRF_UART_EVENT_RXDRDY` while interrupts for `NRF_UART_INT_MASK_RXDRDY` has been disabled (during entry of the `nrfx_uart_rx()` and activated again at exit).

dpgeorge replied:
Ok, I see.  So should `nrfx_uart_rx_enable()` go before or after this line here?  I guess more it before?

glennrub replied:
I'm not sure how many times i have dry-run this scenario in my head by now, but the conclusion is that the current code is the most correct. No re-order needed. 

An attempt on reasoning:

If done the other way around, `nrfx_uart_rx_enable()` before `nrf_uart_rx()`, we might get bytes into the UART HW buffers triggering an IRQ, but as there are no rx-buffer set at this point and it will result in an error instead and the bytes will be skipped.

So, next, why keep it. My current conclusion is that we would need `nrfx_uart_rx_enable()` to prevent `nrf_uart_rx()` to clear the `RXDRDY` issued from the callback handler, and prevent the UART IRQ handler to stop the RX flow (the nrfx UART driver IRQ handler expects to see the `RXDRDY` when the UART IRQ is hit in order to issue a callback to the application). In my head this could happen, but i might be wrong of course.

dpgeorge replied:
Ok, I'll leave the order as-is.

glennrub commented on ports/nrf/modules/machine/uart.c:
```diff
@@ -182,8 +202,10 @@ STATIC mp_obj_t machine_hard_uart_make_new(const mp_obj_type_t *type, size_t n_a
     config.p_context = (void *)self;
 
     // Set NULL as callback function to keep it blocking
```
I guess we can delete this comment as it became a bit misleading :)

dpgeorge replied:
Yes!

glennrub commented:
Great @dpgeorge! It looks very good. Two comments, but not big. I have also tested this with a few different targets, and its working as expected :) 

> Also, maybe newer versions of nrfx handle non-blocking UART differently?
I have also tested pca10040 using nrfx v1.7.0 and on a new target not in the port yet, which is EasyDMA only (UARTE). The proposed change seems to work well with newer versions of nrfx.

dpgeorge commented on ports/nrf/modules/machine/uart.c:
```diff
@@ -70,27 +78,39 @@ STATIC int uart_find(mp_obj_t id) {
-
-    return nrfx_uart_tx(self->p_uart, (uint8_t *)&c, 1);
+    tx_buf[0] = c;
+    tx_pending = true;
```
The `nrfx_uart_tx_in_progress()` call seems to do what I want here but it didn't work, it always blocked.  So that's why I added the `tx_pending` flag and busy wait on it (instead of calling `nrfx_uart_tx_in_progress()`).

glennrub replied:
Yeah, this one was an interesting one. I replaced your code back to `nrfx_uart_tx_in_progress()` and reproduced the issue. However, compiling the code with `DEBUG=1` and` LTO=0` made it work. This hints on a `volatile` bug. So, i updated the `uart_control_block_t.tx_buffer_length` to become a volatile, then it started to work. I'll follow up on the nrfx issue, meanwhile that means your code here is correct as you have moved this logic to micropython and kept it as a volatile. :)

dpgeorge replied:
Good old LTO...

glennrub replied:
The fix i sent to nrfx regarding `nrf_uart_tx_in_progress()` is now in. I have created the PR #4802 upgrading the nrfx to include the fix. It should by this new fix be possible to run the function without getting blocked.

dpgeorge replied:
Great, thanks.  I changed it back to use `nrfx_uart_tx_in_progress` and indeed it works.

dpgeorge commented on ports/nrf/mpconfigport.h:
```diff
@@ -47,7 +47,7 @@
 #define MICROPY_REPL_EMACS_KEYS     (0)
 #define MICROPY_REPL_AUTO_INDENT    (1)
-#define MICROPY_KBD_EXCEPTION       (0)
+#define MICROPY_KBD_EXCEPTION       (1)
```
I think we should enable this by default for all boards.

glennrub replied:
Yes, that makes sense.

dpgeorge commented:
Thanks @glennrub for the feedback.  I've cleaned this PR up and merged it in 302ffdba7f06e8827532834b5902406cd7bf4ea0, with follow up in 456c89f7497d55e641e4374a1860ba08f3b84688

And in 9cf1cbb0575387d79b496b6ff14736ca9239c0a1 I added a call to `wfi` while waiting at the UART REPL to save power, saves about 4mA (on microbit and pca10040).
